mat64/svd.go

```diff
@@ -6,474 +6,182 @@
 package mat64
 
 import (
-	"math"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+	"github.com/gonum/lapack/lapack64"
+	"github.com/gonum/matrix"
 )
 
-type SVDFactors struct {
-	U     *Dense
-	Sigma []float64
-	V     *Dense
-	m, n  int
+// SVDKind specifies the treatment of singular vectors during the
+// factorization.
+// NOTE: Should these be in matrix?
+type SVDKind int
+
+const (
+	// SVDNone specifies that no singular vectors should be computed during
+	// the decomposition.
+	SVDNone SVDKind = iota + 1
+	// SVDThin computes the thin singular vectors, that is, it computes
+	//  A = U~ * Σ * V~^T
+	// where U~ is of size m×min(m,n), Σ is a diagonal matrix of size min(m,n)×min(m,n)
+	// and V~ is of size n×min(m,n).
+	SVDThin
+	// SVDFull computes the full singular value decomposition,
+	//  A = U * Σ * V^T
+	// where U is of size m×m, Σ is an m×n diagonal matrix, and V is an n×n matrix.
+	SVDFull
+)
+
+// SVD is a type for creating and using the Singular Value Decomposition (SVD)
+// of a matrix.
+type SVD struct {
+	kind SVDKind
+
+	s  []float64
+	u  blas64.General
+	vt blas64.General
 }
 
-// SVD performs singular value decomposition for an m-by-n matrix a. The
-// singular value decomposition is an m-by-n orthogonal matrix u, an n-by-n
-// diagonal matrix s, and an n-by-n orthogonal matrix v so that a = u*s*v'. If
-// a is a wide matrix a copy of its transpose is allocated, otherwise a is
-// overwritten during the decomposition. Matrices u and v are only created when
-// wantu and wantv are true respectively.
+// Factorize computes the singular value decomposition (SVD) of the input matrix
+// A. The singular values of A are computed in all cases, while the singular
+// vectors are optionally computed depending on the input kind.
 //
-// The singular values, sigma[k] = s[k][k], are ordered so that
+// The full singular value decomposition (kind == SVDFull) deconstructs A as
+//  A = U * Σ * V^T
+// where Σ is an m×n diagonal matrix of singular vectors, U is an m×m unitary
+// matrix of left singular vectors, and V is an n×n matrix of right singular vectors.
 //
-//  sigma[0] >= sigma[1] >= ... >= sigma[n-1].
+// It is frequently not necessary to compute the full SVD. Computation time and
+// storage costs can be reduced using the appropriate kind. Only the singular
+// values can be computed (kind == SVDNone), or a "thin" representation of the
+// singular values (kind = SVDThin). The thin representation can save a significant
+// amount of memory if m >> n. See the documentation for more information.
 //
-// The matrix condition number and the effective numerical rank can be computed from
-// this decomposition.
-func SVD(a *Dense, epsilon, small float64, wantu, wantv bool) SVDFactors {
+// Factorize returns whether the decomposition succeeded. If the decomposition
+// failed, routines that require a successful factorization are not available
+// (Values, UFromSVD, etc.).
```

Comment:
Don't list them here, but do state that Factorise must have been successful or the method will panic at each method.